lib/utils/circularityUtils.ts

```diff
@@ -0,0 +1,130 @@
+import {
+  ContentItemElementsIndexer,
+  Elements,
+  IContentItem,
+} from "@kontent-ai/delivery-sdk";
+
+/**
+ * Module responsible for detecting and removing circular references
+ * in content items from the Kontent.ai Delivery SDK.
+ */
+
+export type CircularReferenceInfo = {
+  parentItemCodename: string;
+  elementName: string;
+  cycle: string[];
+};
+
+export type ItemCircularReferenceMap = Record<string, CircularReferenceInfo[]>;
+
+type LinkableElement = Elements.LinkedItemsElement | Elements.RichTextElement;
+
+/**
+ * Identifies whether a given element can contain links to other items.
+ * @param obj - The element to check.
+ * @returns True if the element can contain links, false otherwise.
+ */
+const isLinkableElement = (
+  obj: ContentItemElementsIndexer
+): obj is LinkableElement => {
+  return (
+    "linkedItems" in obj && Array.isArray((obj as LinkableElement).linkedItems)
+  );
+};
+
+/**
+ * Performs a depth-first traversal to sanitize any content item references that would create a cycle.
+ * Collects information on identified circular references.
+ * @param data - The content item to sanitize.
+ * @param elementName - The name of the current element being processed.
+ * @param seenCodenames - Set of content item codenames visited during the recursion.
+ * @param circularReferences - Accumulated list of detected circular references.
+ * @returns A tuple containing the sanitized content item and the accumulated circular references.
+ */
+const sanitizeCircularDataInternal = <TItem extends IContentItem>(
+  data: TItem,
+  elementName = "",
+  seenCodenames = new Set<string>(),
+  circularReferences: CircularReferenceInfo[] = []
+): [TItem, CircularReferenceInfo[]] => {
+  if (seenCodenames.has(data.system.codename)) {
+    const seenCodenamesArray = Array.from(seenCodenames);
+    const itemCodename = seenCodenamesArray[seenCodenamesArray.length - 1]!;
+    return [
+      // Replaces circular reference with null. Cast necessary to adhere to type signature.
+      null as unknown as TItem,
+      [
+        ...circularReferences,
```

Comment:
If you'd return just the newfound reference (not all the references that the caller knows already anyway), you wouldn't need to filter all the old references out on line `88`.  Btw. the filtering itself is of quadratic complexity run for each linked item element of all items.

Reply:
good point, I somehow managed to make this part more complicated while trying to simplify the rest ... refactored accordingly